### PR TITLE
Feature: upgrade the cluster

### PR DESCRIPTION
**Pull request by mond77**
solve issue #1944 .
the upgrade process contains :
1. upgrade kubeadm,kubeclt,kubelet.
2. upgrade static pods including api-server, controller-manager,scheduler,etcd.
usage in the docs.

### CODE REVIEW

File: pkg/runtime/defaults/default_kubeadm_config.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ apiServer:
   #    - 10.103.97.2
   extraArgs:
     #    etcd-servers: https://192.168.2.110:2379
-    feature-gates: TTLAfterFinished=true,EphemeralContainers=true
```
Comment (mond77) [inline pkg/runtime/defaults/default_kubeadm_config.go:44]: v1.25 unsuppported

File: pkg/runtime/defaults/default_kubeadm_config.go [hunk 2]
```diff
@@ -66,8 +66,7 @@ apiServer:
 controllerManager:
   extraArgs:
     bind-address: 0.0.0.0
-    feature-gates: TTLAfterFinished=true,EphemeralContainers=true
-    experimental-cluster-signing-duration: 876000h
```
Comment (mond77) [inline pkg/runtime/defaults/default_kubeadm_config.go:70]: v1.25 unsupported

File: pkg/runtime/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		err = k.sshCmdAsync(ip, fmt.Sprintf(uncordonNodeCmd, nodename))
+		now := time.Now()
+		//restart api-server takes a few seconds
+		for err != nil {
+			if time.Now().After(now.Add(2 * time.Minute)) {
+				return err
+			}
+			time.Sleep(5 * time.Second)
+			err = k.sshCmdAsync(ip, fmt.Sprintf(uncordonNodeCmd, nodename))
+		}
```
Comment (mond77) [inline pkg/runtime/upgrade.go:101]: when kubectl connection failure happens, retry the Cmd util api-server restart.

**Comment by mond77** [comment]
tested that Kubernetes:v1.24.0 can upgrade to v1.25.0.

File: pkg/runtime/defaults/default_kubeadm_config.go (hunk 1 above)
Comment (cuisongliu) [reply pkg/runtime/defaults/default_kubeadm_config.go:44]: v125 config delete this config

File: pkg/apply/processor/install.go [hunk 4]
```diff
@@ -160,6 +163,30 @@ func (c *InstallProcessor) PreProcess(cluster *v2.Cluster) error {
...
+
+func (c *InstallProcessor) UpgradeIfNeed(cluster *v2.Cluster) error {
+	logger.Info("Executing UpgradeIfNeed Pipeline in InstallProcessor")
+	for i := range c.NewMounts {
+		img := c.NewMounts[i]
+		if img.Type != v2.RootfsImage {
+			continue
+		}
+		logger.Debug("try Upgrade Cluster to %s", img.Labels["version"])
+		err := c.Runtime.UpgradeCluster(img.Labels["version"])
```
Comment (cuisongliu) [inline pkg/apply/processor/install.go:182]: Use a unified constant

File: pkg/runtime/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+)
+
+const (
+	upgradeApplyCmd = "kubeadm upgrade apply %s"
+	upradeNodeCmd   = "kubeadm upgrade node"
+	drainNodeCmd    = "kubectl drain %s --ignore-daemonsets"
+	uncordonNodeCmd = "kubectl uncordon %s"
+	daemonReload    = "systemctl daemon-reload"
+	restartKubelet  = "systemctl restart kubelet"
+	kubeBinaryPath  = "/var/lib/sealos/data/default/rootfs/bin"
```
Comment (cuisongliu) [inline pkg/runtime/upgrade.go:31]: default here is a variable parameter.

File: pkg/apply/processor/install.go [hunk 6]
```diff
@@ -160,6 +164,30 @@ func (c *InstallProcessor) PreProcess(cluster *v2.Cluster) error {
 			c.NewMounts = append(c.NewMounts, *mount)
 		}
 	}
+	runTime, err := runtime.NewDefaultRuntime(cluster, c.ClusterFile.GetKubeadmConfig())
```
Comment (fanux) [inline pkg/apply/processor/install.go:167]: runTime->runtime

File: pkg/apply/processor/install.go [hunk 7]
```diff
@@ -160,6 +164,30 @@ func (c *InstallProcessor) PreProcess(cluster *v2.Cluster) error {
...
+	if err != nil {
+		return fmt.Errorf("failed to init runtime, %v", err)
+	}
+	c.Runtime = runTime
```
Comment (fanux) [inline pkg/apply/processor/install.go:171]: runtime

File: pkg/apply/processor/install.go [hunk 8]
```diff
@@ -160,6 +164,30 @@ func (c *InstallProcessor) PreProcess(cluster *v2.Cluster) error {
...
+	logger.Info("Executing UpgradeIfNeed Pipeline in InstallProcessor")
+	for _, img := range c.NewMounts {
+		version := images.GetKubeVersionFromImage(img)
+		if version == "" {
+			continue
+		}
+		logger.Debug("try Upgrade Cluster to %s", version)
+		err := c.Runtime.UpgradeCluster(version)
+		if err != nil {
+			logger.Info("upgrade cluster failure")
```
Comment (fanux) [inline pkg/apply/processor/install.go:185]: failure->failed

Comment (fanux) [reply pkg/apply/processor/install.go:185]: Or failed to upgrade cluster

File: pkg/runtime/runtime.go [hunk 9]
```diff
@@ -142,3 +144,18 @@ func (k *KubeadmRuntime) Validate() error {
 	}
 	return nil
 }
+
+func (k *KubeadmRuntime) UpgradeCluster(version string) error {
+	curversion := k.getKubeVersionFromImage()
+	if curversion == version {
+		logger.Info("the cluster version no change")
```
Comment (fanux) [inline pkg/runtime/runtime.go:151]: The cluster version has not changed

**Review by fanux — APPROVED** [review]

**Comment by fanux** [comment]
Add a "how to upgrade cluster" docs in examples or user guide.

**Comment by mond77** [comment]
test: 
1. sealos run : `sealos run labring/kubernetes:v1.24.0 labring/calico:v3.22.1 --masters 192.168.64.8 `
```
root@lmaster:~# kubectl get nodes
NAME      STATUS   ROLES           AGE   VERSION
lmaster   Ready    control-plane   67s   v1.24.0
node1     Ready    <none>          31s   v1.24.0
```
2. upgrade : `sealos run labring/kubernetes:v1.25.0`
```
root@lmaster:~# kubectl get nodes
NAME      STATUS   ROLES           AGE     VERSION
lmaster   Ready    control-plane   7m19s   v1.25.0
node1     Ready    <none>          6m43s   v1.25.0
```
3. scale with new version : `sealos add --masters 192.168.64.9 --nodes 192.168.64.10`
```
root@lmaster:~# kubectl get nodes
NAME       STATUS   ROLES           AGE    VERSION
lmaster    Ready    control-plane   10m    v1.25.0
lmaster1   Ready    control-plane   112s   v1.25.0
node1      Ready    <none>          10m    v1.25.0
node2      Ready    <none>          83s    v1.25.0
```
see details at 
[upgradeTest.txt](https://github.com/labring/sealos/files/10370310/upgradeTest.txt)

**Review by fanux — APPROVED** [review]

**Pull request merged**